[oss] Add version to metadata

ankitageorge opened the pull request:
Summary: We want to add versioning to DCP to the metadata so that whenever planner logic changes, we can use the version on save to determine how to load the data

Test Plan:
added a test

Rollback Plan:

Differential Revision: D76135887




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

Skylion007 commented on torch/distributed/checkpoint/default_planner.py:
```diff
@@ -64,6 +64,7 @@
     "create_default_global_save_plan",
 ]
 
+CURRENT_DCP_VERSION: float = 1.0
```
```suggestion
CURRENT_DCP_VERSION: Final[Literal[float]] = 1.0
```
Types it as close to a constant as you can get in Python.

saumishr commented on torch/distributed/checkpoint/metadata.py:
```diff
@@ -147,6 +147,7 @@ class Metadata:
     planner_data: Any = None
     storage_data: Any = None
     storage_meta: Optional[StorageMeta] = None
+    version: Optional[float] = None
```
We can also run into the situations when the planner and metadata remain the same but we change how I/O handles the reading/writing leading to a different format. Basically planner and storage components are not tightly coupled. To me it seems like, we need component versioning here. Planner and Storage components will provide their own versions. Lets discuss the approach more offline.

XuehaiPan commented on torch/distributed/checkpoint/filesystem.py:
```diff
@@ -94,6 +94,8 @@ class SerializationFormat(Enum):
 
 DEFAULT_SUFFIX = ".distcp"
 
+CURRENT_DEFAULT_WRITER_DCP_VERSION: Final[float] = 1.0
```
The best practice for version number is to use `int` or `str` instead of `float`.

```python
num = 10000 * major + 100 * minor + micro
# e.g. 1.0.0 -> 10000
#      1.2.3 -> 10203
#     2.13.1 -> 21301
```

ankitageorge committed:
[oss] Add version to metadata (#155343)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/155343

We want to add versioning to DCP to the metadata so that whenever planner logic changes, we can use the version on save to determine how to load the data

Test Plan:
added a test

Rollback Plan:

Differential Revision: D76135887

teja-rao reviewed (approved):
Looks good! Thanks for working on this
